beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -0,0 +1,302 @@
+func (f *ForkChoice) AncestorRoot(ctx context.Context, root [32]byte, slot types.Slot) ([]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "protoArray.AncestorRoot")
+	defer span.End()
+
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	node, ok := f.store.nodeByRoot[root]
+	if !ok || node == nil {
+		return nil, errNilNode
+	}
+
+	n := node
+	for n != nil && n.slot > slot {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		n = n.parent
+	}
+
+	if n == nil {
+		return nil, errNilNode
+	}
+
+	return n.root[:], nil
+}
+
+// updateBalances updates the balances that directly voted for each block taking into account the
+// validators' latest votes.
+func (f *ForkChoice) updateBalances(newBalances []uint64) error {
+	for index, vote := range f.votes {
+		// Skip if validator has never voted for current root and next root (i.e. if the
+		// votes are zero hash aka genesis block), there's nothing to compute.
+		if vote.currentRoot == params.BeaconConfig().ZeroHash && vote.nextRoot == params.BeaconConfig().ZeroHash {
+			continue
+		}
+
+		oldBalance := uint64(0)
+		newBalance := uint64(0)
+		// If the validator index did not exist in `f.balances` or
+		// `newBalances` list above, the balance is just 0.
+		if index < len(f.balances) {
+			oldBalance = f.balances[index]
+		}
+		if index < len(newBalances) {
+			newBalance = newBalances[index]
+		}
+
+		// Update only if the validator's balance or vote has changed.
+		if vote.currentRoot != vote.nextRoot || oldBalance != newBalance {
+			// Ignore the vote if the root is not in fork choice
+			// store, that means we have not seen the block before.
+			nextNode, ok := f.store.nodeByRoot[vote.nextRoot]
+			if ok && vote.nextRoot != params.BeaconConfig().ZeroHash {
+				// Protection against nil node
+				if nextNode == nil {
+					return errNilNode
+				}
+				nextNode.balance += newBalance
+			}
+
+			currentNode, ok := f.store.nodeByRoot[vote.currentRoot]
+			if ok && vote.currentRoot != params.BeaconConfig().ZeroHash {
+				// Protection against nil node
+				if currentNode == nil {
+					return errNilNode
+				}
+				if currentNode.balance < oldBalance {
+					return errInvalidBalance
+				}
+				currentNode.balance -= oldBalance
+			}
+		}
+
+		// Rotate the validator vote.
+		f.votes[index].currentRoot = vote.nextRoot
+	}
+	f.balances = newBalances
+	return nil
+}
+
+// Tips returns a list of possible heads from fork choice store, it returns the
+// roots and the slots of the leaf nodes.
+func (f *ForkChoice) Tips() ([][32]byte, []types.Slot) {
+	return f.store.tips()
+}
+
+// ProposerBoost returns the proposerBoost of the store
+func (f *ForkChoice) ProposerBoost() [fieldparams.RootLength]byte {
+	return f.store.proposerBoost()
+}
+
+// ConfirmOptimisticValidNode sets the node with the given root as a fully validated node
+func (f *ForkChoice) ConfirmOptimisticValidNode(ctx context.Context, root [fieldparams.RootLength]byte) error {
```

Comment:
It is a little too mouthful. "Confirm" may not be the right verb here.

How about something like `SetNodeValid`, `UpdateValidNode`, `UpdateNodeToValid`... etc ?